File: src/kernel/extensions/registry.py
"""In-memory registries for extension and plugin manifests."""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass
from typing import List, Literal

from src.kernel.extensions.manifest import (
    ExtensionManifest,
    ExtensionType,
    PluginLifecycleHook,
    PluginManifest,
    PluginRoute,
    PluginTool,
)

LifecyclePhase = Literal["startup", "shutdown"]


class RegistryDuplicateError(ValueError):
    """Raised when a registry receives the same manifest id twice."""


@dataclass(frozen=True)
class PluginLifecycleHookRegistration:
    """A lifecycle hook declaration with registry ordering metadata."""

    plugin_id: str
    hook: PluginLifecycleHook
    plugin_index: int
    hook_index: int

    @property
    def name(self) -> str:
        return self.hook.name

    @property
    def module(self) -> str:
        return self.hook.module

    @property
    def phase(self) -> LifecyclePhase:
        return self.hook.phase

    @property
    def order(self) -> int:
        return self.hook.order


@dataclass(frozen=True)
class PluginRouteRegistration:
    """A backend route declaration with plugin ownership metadata."""

    plugin_id: str
    route: PluginRoute

    @property
    def name(self) -> str:
        return self.route.name

    @property
    def prefix(self) -> str:
        return self.route.prefix

    @property
    def module(self) -> str:
        return self.route.module

    @property
    def required_permissions(self) -> list[str]:
        return self.route.required_permissions

    @property
    def tags(self) -> list[str]:
        return self.route.tags


@dataclass(frozen=True)
class PluginToolRegistration:
    """A tool declaration with plugin ownership metadata."""

    plugin_id: str
    tool: PluginTool

    @property
    def name(self) -> str:
        return self.tool.name

    @property
    def module(self) -> str:
        return self.tool.module

    @property
    def required_permissions(self) -> list[str]:
        return self.tool.required_permissions

    @property
    def legacy_ids(self) -> list[str]:
        return self.tool.legacy_ids


class ExtensionRegistry:
    """Build-time registry for all extension manifests."""

    def __init__(self, manifests: Iterable[ExtensionManifest] | None = None) -> None:
        self._items: dict[str, ExtensionManifest] = {}
        if manifests:
            for manifest in manifests:
                self.register(manifest)

    def register(self, manifest: ExtensionManifest) -> ExtensionManifest:
        if manifest.id in self._items:
            raise RegistryDuplicateError(f"extension already registered: {manifest.id}")
        self._items[manifest.id] = manifest
        return manifest

    def get(self, extension_id: str) -> ExtensionManifest | None:
        return self._items.get(extension_id)

    def list(
        self,
        *,
        extension_type: ExtensionType | str | None = None,
        enabled: bool | None = None,
    ) -> list[ExtensionManifest]:
        items = list(self._items.values())
        if extension_type is not None:
            expected = ExtensionType(extension_type)
            items = [item for item in items if item.type == expected]
        if enabled is not None:
            items = [item for item in items if item.enabled is enabled]
        return items

    def permissions(self, *, enabled_only: bool = True) -> List[str]:
        seen: set[str] = set()
        result: List[str] = []
        for manifest in self._items.values():
            if enabled_only and not manifest.enabled:
                continue
            for permission in manifest.permissions:
                if permission not in seen:
                    seen.add(permission)
                    result.append(permission)
        return result

class PluginRegistry:
    """Build-time registry for plugin manifests."""

    def __init__(self, manifests: Iterable[PluginManifest] | None = None) -> None:
        self._items: dict[str, PluginManifest] = {}
        if manifests:
            for manifest in manifests:
                self.register(manifest)

    def register(self, manifest: PluginManifest) -> PluginManifest:
        if manifest.id in self._items:
            raise RegistryDuplicateError(f"plugin already registered: {manifest.id}")
        self._items[manifest.id] = manifest
        return manifest

    def get(self, plugin_id: str) -> PluginManifest | None:
        return self._items.get(plugin_id)

    def list(self, *, enabled_by_default: bool | None = None) -> list[PluginManifest]:
        items = list(self._items.values())
        if enabled_by_default is not None:
            items = [item for item in items if item.enabled_by_default is enabled_by_default]
        return items

    def as_extension_registry(self, *, publisher: str = "core") -> ExtensionRegistry:
        return ExtensionRegistry(
            manifest.as_extension_manifest(publisher=publisher) for manifest in self._items.values()
        )

    def permissions(self, *, enabled_only: bool = True) -> List[str]:
        seen: set[str] = set()
        result: List[str] = []
        for manifest in self._items.values():
            if enabled_only and not manifest.enabled_by_default:
                continue
            for permission in manifest.declared_permissions():
                if permission not in seen:
                    seen.add(permission)
                    result.append(permission)
        return result

    def routes(self, *, enabled_only: bool = True) -> List[PluginRouteRegistration]:
        registrations: List[PluginRouteRegistration] = []
        for manifest in self._items.values():
            if enabled_only and not manifest.enabled_by_default:
                continue
            for route in manifest.routers:
                registrations.append(
                    PluginRouteRegistration(plugin_id=manifest.id, route=route)
                )
        return registrations

    def tools(self, *, enabled_only: bool = True) -> List[PluginToolRegistration]:
        registrations: List[PluginToolRegistration] = []
        for manifest in self._items.values():
            if enabled_only and not manifest.enabled_by_default:
                continue
            for tool in manifest.tools:
                registrations.append(
                    PluginToolRegistration(plugin_id=manifest.id, tool=tool)
                )
        return registrations

    def lifecycle_hooks(
        self,
        *,
        phase: LifecyclePhase | None = None,
        enabled_only: bool = True,
    ) -> List[PluginLifecycleHookRegistration]:
        """Return lifecycle hook declarations in stable execution order.

        This method only collects declarations. It intentionally does not import
        hook modules or execute hook callables; the future execution layer must
        isolate import/runtime errors so core startup and shutdown are not blocked.
        """
        registrations: List[PluginLifecycleHookRegistration] = []
        for plugin_index, manifest in enumerate(self._items.values()):
            if enabled_only and not manifest.enabled_by_default:
                continue
            for hook_index, hook in enumerate(manifest.lifespan_hooks):
                if phase is not None and hook.phase != phase:
                    continue
                registrations.append(
                    PluginLifecycleHookRegistration(
                        plugin_id=manifest.id,
                        hook=hook,
                        plugin_index=plugin_index,
                        hook_index=hook_index,
                    )
                )
        return sorted(
            registrations,
            key=lambda item: (item.order, item.plugin_index, item.hook_index),
        )

File: frontend/src/extensions/registry.ts
import type {
  ExtensionManifest,
  ExtensionRegistryListOptions,
  ExtensionType,
  PluginManifest,
  PluginNavItem,
  PluginPanel,
  PluginRegistryListOptions,
  PluginRoute,
} from "./types";

export class RegistryDuplicateError extends Error {
  constructor(kind: "extension" | "plugin", id: string) {
    super(`${kind} already registered: ${id}`);
    this.name = "RegistryDuplicateError";
  }
}

function normalizeStringList(values: readonly string[] | undefined): string[] {
  const seen = new Set<string>();
  const result: string[] = [];
  for (const value of values ?? []) {
    const normalized = value.trim();
    if (normalized && !seen.has(normalized)) {
      seen.add(normalized);
      result.push(normalized);
    }
  }
  return result;
}

function isEnabled(value: boolean | undefined): boolean {
  return value !== false;
}

function contributionListOptions(
  options: PluginRegistryListOptions,
): PluginRegistryListOptions {
  return { enabled: options.enabled ?? true };
}

export function normalizeExtensionManifest(
  manifest: ExtensionManifest,
): ExtensionManifest {
  return {
    ...manifest,
    id: manifest.id.trim(),
    tags: normalizeStringList(manifest.tags),
    capabilities: normalizeStringList(manifest.capabilities),
    permissions: normalizeStringList(manifest.permissions),
    installState: manifest.installState ?? "builtin",
    enabled: isEnabled(manifest.enabled),
  };
}

export function collectPluginPermissions(manifest: PluginManifest): string[] {
  return normalizeStringList([
    ...(manifest.permissions ?? []),
    ...collectContributionPermissions(manifest.routes),
    ...collectContributionPermissions(manifest.panels),
    ...collectContributionPermissions(manifest.navItems),
  ]);
}

function collectContributionPermissions(
  contributions: Array<{ requiredPermissions?: string[] }> | undefined,
): string[] {
  return (contributions ?? []).flatMap(
    (contribution) => contribution.requiredPermissions ?? [],
  );
}

export class ExtensionRegistry {
  private readonly items = new Map<string, ExtensionManifest>();

  constructor(manifests: readonly ExtensionManifest[] = []) {
    for (const manifest of manifests) {
      this.register(manifest);
    }
  }

  register(manifest: ExtensionManifest): ExtensionManifest {
    const normalized = normalizeExtensionManifest(manifest);
    if (!normalized.id) {
      throw new Error("extension id cannot be blank");
    }
    if (this.items.has(normalized.id)) {
      throw new RegistryDuplicateError("extension", normalized.id);
    }
    this.items.set(normalized.id, normalized);
    return normalized;
  }

  get(id: string): ExtensionManifest | undefined {
    return this.items.get(id);
  }

  list(options: ExtensionRegistryListOptions = {}): ExtensionManifest[] {
    return Array.from(this.items.values()).filter((manifest) => {
      if (options.type && manifest.type !== options.type) {
        return false;
      }
      if (options.enabled !== undefined && isEnabled(manifest.enabled) !== options.enabled) {
        return false;
      }
      return true;
    });
  }

  permissions(options: { enabledOnly?: boolean } = {}): string[] {
    const enabledOnly = options.enabledOnly ?? true;
    return normalizeStringList(
      this.list()
        .filter((manifest) => !enabledOnly || isEnabled(manifest.enabled))
        .flatMap((manifest) => manifest.permissions ?? []),
    );
  }
}

export class PluginRegistry {
  private readonly items = new Map<string, PluginManifest>();

  constructor(manifests: readonly PluginManifest[] = []) {
    for (const manifest of manifests) {
      this.register(manifest);
    }
  }

  register(manifest: PluginManifest): PluginManifest {
    const id = manifest.id.trim();
    if (!id) {
      throw new Error("plugin id cannot be blank");
    }
    if (this.items.has(id)) {
      throw new RegistryDuplicateError("plugin", id);
    }
    const normalized = { ...manifest, id };
    this.items.set(id, normalized);
    return normalized;
  }

  get(id: string): PluginManifest | undefined {
    return this.items.get(id);
  }

  list(options: PluginRegistryListOptions = {}): PluginManifest[] {
    return Array.from(this.items.values()).filter((manifest) => {
      if (options.enabled !== undefined) {
        return isEnabled(manifest.enabledByDefault) === options.enabled;
      }
      return true;
    });
  }

  routes(options: PluginRegistryListOptions = {}): PluginRoute[] {
    return this.list(contributionListOptions(options)).flatMap((manifest) =>
      (manifest.routes ?? []).filter((route) => isEnabled(route.enabled)),
    );
  }

  panels(options: PluginRegistryListOptions = {}): PluginPanel[] {
    return this.list(contributionListOptions(options)).flatMap((manifest) =>
      (manifest.panels ?? []).filter((panel) => isEnabled(panel.enabled)),
    );
  }

  navItems(options: PluginRegistryListOptions = {}): PluginNavItem[] {
    return this.list(contributionListOptions(options))
      .flatMap((manifest) =>
        (manifest.navItems ?? []).filter((item) => isEnabled(item.enabled)),
      )
      .sort((left, right) => (left.order ?? 100) - (right.order ?? 100));
  }

  settingsSections(options: PluginRegistryListOptions = {}): string[] {
    return normalizeStringList(
      this.list(contributionListOptions(options)).flatMap(
        (manifest) => manifest.settingsSections ?? [],
      ),
    );
  }

  toolRenderers(options: PluginRegistryListOptions = {}): string[] {
    return normalizeStringList(
      this.list(contributionListOptions(options)).flatMap(
        (manifest) => manifest.toolRenderers ?? [],
      ),
    );
  }

  i18nNamespaces(options: PluginRegistryListOptions = {}): string[] {
    return normalizeStringList(
      this.list(contributionListOptions(options)).flatMap(
        (manifest) => manifest.i18nNamespaces ?? [],
      ),
    );
  }

  permissions(options: { enabledOnly?: boolean } = {}): string[] {
    const enabledOnly = options.enabledOnly ?? true;
    return normalizeStringList(
      this.list({ enabled: enabledOnly ? true : undefined }).flatMap((manifest) =>
        collectPluginPermissions(manifest),
      ),
    );
  }

  asExtensionRegistry(options: { publisher?: string } = {}): ExtensionRegistry {
    const publisher = options.publisher ?? "core";
    return new ExtensionRegistry(
      this.list().map((manifest) => ({
        id: manifest.id,
        type: "plugin" as ExtensionType,
        name: manifest.name,
        version: manifest.version,
        publisher,
        capabilities: ["plugin"],
        permissions: collectPluginPermissions(manifest),
        installState: manifest.core ? "builtin" : "installed",
        enabled: isEnabled(manifest.enabledByDefault),
        compatibility: { apiVersion: manifest.apiVersion },
      })),
    );
  }
}
